### ISSUE TITLE
Terraform init fails with checkpoint error

### ISSUE THREAD

**Issue by pradeepbhadani** [hashicorp/terraform#26716]
### Terraform Version

```
Terraform v0.13.5
```

#### Terraform version command throws the same error
```
TF_LOG=TRACE terraform --version
```
```
2020/10/27 10:10:46 [INFO] Terraform version: 0.13.5
2020/10/27 10:10:46 [INFO] Go runtime version: go1.14.7
2020/10/27 10:10:46 [INFO] CLI args: []string{"/Users/username/tools/terraform/terraform", "--version"}
2020/10/27 10:10:46 [DEBUG] Attempting to open CLI config file: /Users/username/.terraformrc
2020/10/27 10:10:46 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/10/27 10:10:46 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2020/10/27 10:10:46 [DEBUG] ignoring non-existing provider search directory /Users/username/.terraform.d/plugins
2020/10/27 10:10:46 [DEBUG] ignoring non-existing provider search directory /Users/username/Library/Application Support/io.terraform/plugins
2020/10/27 10:10:46 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2020/10/27 10:10:46 [INFO] CLI command args: []string{"version", "--version"}
2020/10/27 10:10:49 [ERR] Checkpoint error: Get "https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin&signature=8d5ZZZZ-YYYY-b2c0-c3a1-XXXXXdfb5&version=0.13.5": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Terraform v0.13.5
```

#### Terraform init throws following error
```
TF_LOG=TRACE terraform init
```
```
2020/10/27 10:04:12 [INFO] Terraform version: 0.13.5
2020/10/27 10:04:12 [INFO] Go runtime version: go1.14.7
2020/10/27 10:04:12 [INFO] CLI args: []string{"/Users/username/tools/terraform/terraform", "init"}
2020/10/27 10:04:12 [DEBUG] Attempting to open CLI config file: /Users/username/.terraformrc
2020/10/27 10:04:12 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/10/27 10:04:12 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2020/10/27 10:04:12 [DEBUG] ignoring non-existing provider search directory /Users/username/.terraform.d/plugins
2020/10/27 10:04:12 [DEBUG] ignoring non-existing provider search directory /Users/username/Library/Application Support/io.terraform/plugins
2020/10/27 10:04:12 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2020/10/27 10:04:12 [INFO] CLI command args: []string{"init"}
2020/10/27 10:04:12 [TRACE] ModuleInstaller: installing child modules for . into .terraform/modules
Initializing modules...
2020/10/27 10:04:12 [DEBUG] Module installer: begin raw-bucket
2020/10/27 10:04:12 [TRACE] ModuleInstaller: Module installer: raw-bucket <nil> already installed in .terraform/modules/raw-bucket
2020/10/27 10:04:12 [TRACE] modsdir: writing modules manifest to .terraform/modules/modules.json

Initializing the backend...
2020/10/27 10:04:12 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2020/10/27 10:04:12 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2020/10/27 10:04:12 [DEBUG] New state was assigned lineage "c01xxxxx-zzzz-0000-14f9-qqqqq33335f"
2020/10/27 10:04:12 [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2020/10/27 10:04:12 [TRACE] Meta.Backend: instantiated backend of type <nil>
2020/10/27 10:04:12 [DEBUG] checking for provisioner in "."
2020/10/27 10:04:12 [DEBUG] checking for provisioner in "/Users/username/tools/terraform"
2020/10/27 10:04:12 [INFO] Failed to read plugin lock file .terraform/plugins/darwin_amd64/lock.json: open .terraform/plugins/darwin_amd64/lock.json: no such file or directory
2020/10/27 10:04:12 [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend

Initializing provider plugins...
2020/10/27 10:04:12 [TRACE] backend/local: state manager for workspace "default" will:
 - read initial snapshot from terraform.tfstate
 - write new snapshots to terraform.tfstate
 - create any backup at terraform.tfstate.backup
2020/10/27 10:04:12 [TRACE] statemgr.Filesystem: reading initial snapshot from terraform.tfstate
2020/10/27 10:04:12 [TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay
2020/10/27 10:04:12 [TRACE] statemgr.Filesystem: read nil snapshot
2020/10/27 10:04:12 [TRACE] providercache.fillMetaCache: scanning directory .terraform/plugins
2020/10/27 10:04:12 [TRACE] getproviders.SearchLocalDirectory: .terraform/plugins is a symlink to .terraform/plugins
2020/10/27 10:04:12 [DEBUG] Service discovery for registry.terraform.io at https://registry.terraform.io/.well-known/terraform.json
2020/10/27 10:04:12 [TRACE] HTTP client GET request to https://registry.terraform.io/.well-known/terraform.json
- Finding hashicorp/aws versions matching "~> 3.0"...
2020/10/27 10:04:15 [ERR] Checkpoint error: Get "https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin&signature=8d5ZZZZZ-YYYY-b2c0-XXXX-exxxxx7dfb5&version=0.13.5": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/aws:
could not connect to registry.terraform.io: Failed to request discovery
document: Get "https://registry.terraform.io/.well-known/terraform.json":
context deadline exceeded (Client.Timeout exceeded while awaiting headers)

```

I tried to set following variables and still the same error
```
export TF_REGISTRY_CLIENT_TIMEOUT=120
```
```
 export CHECKPOINT_TIMEOUT=50000
```

### Expected Behavior
Terraform init should download the provider.

### Actual Behavior
Terraform init fails to download provider.

### Steps to Reproduce
`terraform init` and `terraform version` throws similar error.

### Additional Information
1. Curl command successfully resolve the terraform endpoints.
```
curl "https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin&version=0.13.5"
{"product":"terraform","current_version":"0.13.5","current_release":1603310232,"current_download_url":"https://releases.hashicorp.com/terraform/0.13.5/","current_changelog_url":"https://github.com/hashicorp/terraform/blob/v0.13.5/CHANGELOG.md","project_website":"https://www.terraform.io","alerts":[]}%
```

2. Able to resolve the endpoint using `go` program. Although it takes couple of second to run and return result.

3. Facing same issue with `terraform v0.13.4` and `terraform v0.13.5`.

### References
- #26086

**Comment by mildwonkey** [comment]
@pradeepbhadani , I am sorry you are experiencing this error!  This issue looks pretty similar: https://github.com/hashicorp/terraform/issues/26532

What OS are you on - I can see it's a mac, but what specific version please? Do you have any custom proxy or DNS settings?

**Comment by pradeepbhadani** [comment]
@mildwonkey I am on MacOS Catalina v10.15.7
I am using a corporate MacBook so might have proxy settings. I don't see any specific settings for DNS.
But I am able to resolve the hashicorp registry and other endpoints using `curl` and a `go` program.

I removed `~/.terraform.d`, `.terraform` and updated my DNS to use `8.8.8.8` and getting the similar error:

```
 TF_LOG=TRACE terraform version
2020/10/29 12:16:04 [INFO] Terraform version: 0.13.5
2020/10/29 12:16:04 [INFO] Go runtime version: go1.14.7
2020/10/29 12:16:04 [INFO] CLI args: []string{"/Users/pbhadani/tools/terraform/terraform", "version"}
2020/10/29 12:16:04 [DEBUG] Attempting to open CLI config file: /Users/pbhadani/.terraformrc
2020/10/29 12:16:04 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/10/29 12:16:04 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2020/10/29 12:16:04 [DEBUG] ignoring non-existing provider search directory /Users/pbhadani/.terraform.d/plugins
2020/10/29 12:16:04 [DEBUG] ignoring non-existing provider search directory /Users/pbhadani/Library/Application Support/io.terraform/plugins
2020/10/29 12:16:04 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2020/10/29 12:16:04 [INFO] CLI command args: []string{"version"}
2020/10/29 12:16:14 [ERR] Checkpoint error: Get "https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin&signature=926f8XXXX-yyyy-0000-zzz-22222224e071a&version=0.13.5": dial tcp: lookup checkpoint-api.hashicorp.com on 8.8.8.8:53: read udp 127.0.0.1:63350->127.0.0.1:39705: i/o timeout
Terraform v0.13.
```

**Comment by mildwonkey** [comment]
The reason I ask about DNS - and the reason that `curl` working doesn't necessarily tell us anything - is that go has a long-standing open issue where any `/etc/resolver` dns settings aren't honored on mac: https://github.com/golang/go/issues/12524

I'm not sure why this is causing new problems for users today; but @paultyng had a good suggestion another open timeout issue. Could you try with 0.14-beta and let us know if that works for you? It is a beta, so please not on production infrastructure! 

https://github.com/hashicorp/terraform/issues/26532#issuecomment-719012852

**Comment by pradeepbhadani** [comment]
@mildwonkey  I tried `tf-0.14-beta-2` , `TF_REGISTRY_CLIENT_TIMEOUT` and `TF_REGISTRY_DISCOVERY_RETRY` but same result.

```
$% TF_REGISTRY_CLIENT_TIMEOUT=120
$% TF_REGISTRY_DISCOVERY_RETRY=5
$% TF_LOG=TRACE /Users/username/tools/tf-14beta/terraform version
2020/11/02 09:15:45 [INFO] Terraform version: 0.14.0 beta2
2020/11/02 09:15:45 [INFO] Go runtime version: go1.15.2
2020/11/02 09:15:45 [INFO] CLI args: []string{"/Users/username/tools/tf-14beta/terraform", "version"}
2020/11/02 09:15:45 [DEBUG] Attempting to open CLI config file: /Users/username/.terraformrc
2020/11/02 09:15:45 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/11/02 09:15:45 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2020/11/02 09:15:45 [DEBUG] ignoring non-existing provider search directory /Users/username/.terraform.d/plugins
2020/11/02 09:15:45 [DEBUG] ignoring non-existing provider search directory /Users/username/Library/Application Support/io.terraform/plugins
2020/11/02 09:15:45 [DEBUG] ignoring non-existing provider search directory /Library/Application Support/io.terraform/plugins
2020/11/02 09:15:45 [INFO] CLI command args: []string{"version"}
2020/11/02 09:15:48 [ERR] Checkpoint error: Get "https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin&signature=0000000-xxxx-yyyy-zzz-qqqqqqqq&version=0.14.0-beta2": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Terraform v0.14.0-beta2
```

Also, `go` code is also working
```
$% cat hashi_registry.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
)

func main() {
	res, err := http.Get("https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin")
	if err != nil {
		log.Fatal(err)
	}
	res_body, err := ioutil.ReadAll(res.Body)
	res.Body.Close()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%s", res_body)
}
```

```
$% go run hashi_registry.go
{"product":"terraform","current_version":"0.13.5","current_release":1603310232,"current_download_url":"https://releases.hashicorp.com/terraform/0.13.5/","current_changelog_url":"https://github.com/hashicorp/terraform/blob/v0.13.5/CHANGELOG.md","project_website":"https://www.terraform.io","alerts":[]}
```

**Comment by mildwonkey** [comment]
@pradeepbhandani thank you very much for trying these for me, I really appreciate it. I was finally able to reproduce this locally, specifically in the case where `terraform init` was failing while curl, dig, etc worked just fine.

This is how I reproduced this on my local mac:

1. modified my DNS settings to use entirely non-existent DNS resolvers (3.3.3.3, 2.2.2.2)
1. manually edited /etc/resolv.conf to use *actual* dns resolvers
    - curl, dig, etc registry.terraform.io: works just fine
    - terraform init: fails w/ timeout
1. re-build terraform locally with CGO_ENABLED=1
    - now it works 

So, as we've suspected for awhile, the issues lies with golang/go#12524. We have reached out to our build team here at hashicorp about re-working our build to not use CGO. 

One caveat: when we make this change, terraform will no longer have this issue, but it's entirely possible that folks will still see this same problem again because individual providers might be built with CGO (`terraform init` would succeed but the actual plan/apply could fail with similar looking errors).

We will try and coordinate this build change with the provider developers we have direct relationships with, and communicate the necessity with all provider developers, but since we don't control their build processes we can't guarantee that all providers will make this same change. (this is not immediately relevant to the issue at hand, but I wanted to get the information out there in for future reference).
